### ISSUE TITLE
Quantity and PnL in Position

### ISSUE THREAD

**Issue by rsmb7z** [nautechsystems/nautilus_trader#933]
# Feature Request

In the `engine.trader.generate_positions_report()` it shows `peak_qty` (which is the peak qty of instrument during run and not of open/close qty) and the `realized_pnl` (which shows cummulative/running total of realized PnL). I think would be useful to have follwoing two fields in Position object (for that specific open/close):

1. PnL
2. Quantity (retain quantity after the position is closed)

Although the actual quantity can be found in orders fill report, but thats not as trade style like for positions report (open/close).

**Comment by cjdsellers** [comment]
So the `peak_quantity` will be the greatest quantity reached by the position between open and close (not per instrument):
```
        if self.quantity._mem.raw > self.peak_qty._mem.raw:
            self.peak_qty = self.quantity
```

As for the PnL, this is just a cumulative calculation when there is a closing trade, just based on the quote currency for the instrument:
```
realized_pnl += self._calculate_pnl(self.avg_px_open, last_px, last_qty)
```

To cover your suggestions:
1. **PnL:** Is this total PnL realized + unrealized? To get unrealized the position needs to know where the market currently is. Because of this there is already a `.total_pnl(Price last)` method.
2. **Quantity:** Is this the initial quantity for the position? I think peak quantity might be what you want, as it may take multiple fills for the position to reach its 'initial' quantity from the first order? or maybe the initial quantity from the first order?

**Comment by rsmb7z** [comment]
> So the `peak_quantity` will be the greatest quantity reached by the position between open and close (not per instrument):
> 
> ```
>         if self.quantity._mem.raw > self.peak_qty._mem.raw:
>             self.peak_qty = self.quantity
> ```

I have attached two snapshots from Fills report and Positions report. Please have a look at `peak_qty` from positions report and actual trade `quantity` from fill report.
The same can be reproduced with any strategy which uses dynamic trade_size like `round(RiskAmount / last_bar.close, 0)` and significant change in price to have different qty in trades.


#### From Fills Report: `engine.trader.generate_order_fills_report()`
![fills](https://user-images.githubusercontent.com/105105941/210041268-9fc22999-c514-452c-8da0-e5a5f260c86a.png)

#### From Positions Report: `engine.trader.generate_positions_report()`
![positions](https://user-images.githubusercontent.com/105105941/210041255-3ffd0a15-a84e-4056-86e0-96c06011580f.png)
(blue highlights are the expected columns to have)


> ```
> realized_pnl += self._calculate_pnl(self.avg_px_open, last_px, last_qty)
> ```

> 1. **PnL:** Is this total PnL realized + unrealized? To get unrealized the position needs to know where the market currently is. Because of this there is already a `.total_pnl(Price last)` method.

Only the realized profit in this case as the report is for closed positions only. While having cummulative as before is fine as well.

> 2. **Quantity:** Is this the initial quantity for the position? I think peak quantity might be what you want, as it may take multiple fills for the position to reach its 'initial' quantity from the first order? or maybe the initial quantity from the first order?

`peak_qty` is fine as-is and as the name suggest. You can see in Positions snapshot `peak_qty` is the maximum quantity traded during the trader session, where in last few Orders actual quantity is different (see from Fills report). `peak_qty` is different for each position (actually broker's trading system no longer call that position once it is closed, instead refer as trade), which is the peak_qty when the position object was added to Cache and is correct behavior.

**Comment by cjdsellers** [comment]
Thanks for the screenshots, it's interesting to see the behavior _in the wild_.

Could the differences in avg_px be accounted for with commissions?

Are there any other action items for this one then? I think everything is behaving as intended?

**Comment by rsmb7z** [comment]
Hi @cjdsellers 

> Could the differences in avg_px be accounted for with commissions?

I have reviewed avg_px_open and avg_px_close and didn't find anything wrong with it so far.
Is the commission config in BacktestEngineConfig? I couldn't find in example to run test with commissions to see affect on pnl, so in my case it is 0 for now.
Ideally the commission shouldn't be part of avg_px and shall remain actual fill price and commission being seperate field, however the `realized_pnl` shall have commission accounted for.

> Are there any other action items for this one then? I think everything is behaving as intended?

I find that `PnL Statistics` at end of Backtest, are depending on 'realized_pnl' (which infact is cummulative) hence wrong results for Max, Min, Avg etc. So `realized_pnl` should be of that position only whereas for running total can be renamed to more appropriate name i.e. realized_pnl_to_date etc.

**Comment by cjdsellers** [comment]
Just to confirm, the statistics using Max, Min, Avg are per `Position`.

So I think everything is correct unless there are specific actionables to address?

**Comment by rsmb7z** [comment]
> Just to confirm, the statistics using Max, Min, Avg are per `Position`.

Max, Min, Avg is of all positions, the way it is now I believe is correct apart from the fact that field/column it is depending on `realized_pnl` should have per position pnl instead of cummulative/running total.

> So I think everything is correct unless there are specific actionables to address?

Yes correct, I didn't find anything else wrong so far.

**Comment by cjdsellers** [comment]
Sure, could you please provide an example of how the `generate_positions_report` produces any cumulative/running totals?

I believe the report is per position, unless you're referring to a different output somewhere?

**Comment by rsmb7z** [comment]
I think I found something new, things are changing based on OmsType. You can reproduce as follows:

- Use `fx_ema_cross_audusd_ticks` from examples\backtest. 
- Run the backtest, everything looks as expected
- Modify to OmsType.NETTING
- Run the backtest again and compare realized_pnl column of generate_positions_report in both.

So OmsType.HEDGING is showing positions realized_pnl whereas OmsType.NETTING making it cummulative. What I understand is the Hedging and Netting is the trading style where later allows you to hold one side of position but the prior allows to hold long/short at same time. Based on this assumption, my test case is set to OmsType.NETTING. Anyways in above tests performance is same in both cases because of same Strategy.

Now coming to reports, in either case shouldn't it be same? As you will see the results, orders/positions etc is same because Strategy is trading one side.

![Capture](https://user-images.githubusercontent.com/105105941/212278184-7b926cbf-d5d8-4fa0-8b09-9953607dd3e3.JPG)

**Comment by cjdsellers** [comment]
So you've actually uncovered a bug. When positions are being generated in `NETTING` OMS mode, where position snapshots are taken by the cache, subsequently bringing them all back from the cache results in `realized_pnl` becoming cumulative across all the snapshot positions (which is not the desired behavior).

We're currently investigating, this attribute seems to be the only one with this behavior, other others e.g. `commissions` are correct per netted position.

**Comment by cjdsellers** [comment]
Now fixed on `develop` from ff0a252ce28d6cce40a7cce2d0992e82ee3ad3e3.

**Comment by rsmb7z** [comment]
Hi @cjdsellers 
Second part of request about `peak_qty` is behaving exactly same as `realized_pnl` was. To reproduce, you will have to modify the strategy slightly to use different quantity during backtest run. Or you can run on any stock and position_size using Investment/price.

**Comment by cjdsellers** [comment]
Thanks again for the report, also fixed that from 7594292592bb1a20dfbd36260b45c9fd3eef6245.

**Comment by OnlyC** [comment]
I built from main + applied 2 above fixed already. The **avg_px_open** looks like doesn't change when reverse position.

2022-08-02T19:37:46 - I'm having a SHORT -1.1 @2.7810247713884997
2022-08-03T01:26:46 - Market BUY 375.8 @2.662. Position changed event show side=LONG, net_qty=374.7 but avg_px_open is the same @2.7810247713884997. This should be @2.662
2022-08-03T03:29:32 - Stop Market BUY 1.1 @2.781 do add to position, avg_px_open is calculated this time.

```
**2022-08-02T19:37:46.320000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] PositionChanged(instrument_id=GTCUSDT-PERP.BINANCE, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, account_id=BINANCE-001, opening_order_id=O-20220802-001-semiauto_ua_rt_dd_market-9, closing_order_id=None, entry=SELL, side=SHORT, net_qty=-1.1, quantity=1.1, peak_qty=367.6, currency=USDT, avg_px_open=2.7810247713884997, avg_px_close=2.7974889510489502, realized_return=-0.00592, realized_pnl=-5.44157789 USDT, unrealized_pnl=-0.00657275 USDT, ts_opened=1659459164621000000, ts_last=1659469066320000000, ts_closed=0, duration_ns=0).**
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: [CMD]--> CancelOrder(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-12, venue_order_id=BINANCE-1-012).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderInitialized(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, side=BUY, type=STOP_MARKET, quantity=1.1, time_in_force=GTC, post_only=False, reduce_only=True, options={'trigger_price': '2.781', 'trigger_type': 'DEFAULT', 'expire_time_ns': 0}, emulation_trigger=NO_TRIGGER, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, tags=semiauto_ua_rt_dd_market,stop_close).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: [CMD]--> SubmitOrder(order=StopMarketOrder(BUY 1.1 GTCUSDT-PERP.BINANCE STOP_MARKET @ 2.781[DEFAULT] GTC, status=INITIALIZED, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, venue_order_id=None, tags=semiauto_ua_rt_dd_market,stop_close), position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, exec_algorithm_spec=None).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderSubmitted(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderPendingCancel(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-12, venue_order_id=BINANCE-1-012, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=99_974.17162303 USDT, locked=2.13631486 USDT, free=99_972.03530817 USDT)], margins=[MarginBalance(initial=2.13320681 USDT, maintenance=0.00310805 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=bd2d13d0-5c38-42c2-8003-5e712858e127).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderCanceled(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-12, venue_order_id=BINANCE-1-012, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=99_974.17162303 USDT, locked=2.14253095 USDT, free=99_972.02909208 USDT)], margins=[MarginBalance(initial=2.13942290 USDT, maintenance=0.00310805 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=cac66b86-e915-4a99-9415-e8630a1a9a46).
2022-08-02T23:52:16.680000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderAccepted(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, venue_order_id=BINANCE-1-015, account_id=BINANCE-001, ts_event=1659484336680000000).
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderInitialized(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, side=BUY, type=MARKET, quantity=375.8, time_in_force=GTC, post_only=False, reduce_only=False, options={}, emulation_trigger=NO_TRIGGER, contingency_type=NO_CONTINGENCY, order_list_id=None, linked_order_ids=None, parent_order_id=None, tags=semiauto_ua_rt_dd_market,market_order,rt).   
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: [CMD]--> SubmitOrder(order=MarketOrder(BUY 375.8 GTCUSDT-PERP.BINANCE MARKET GTC, status=INITIALIZED, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, 
venue_order_id=None, tags=semiauto_ua_rt_dd_market,market_order,rt), position_id=None, exec_algorithm_spec=None).
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderSubmitted(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, account_id=BINANCE-001, ts_event=1659490006778000000).
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.50098028 USDT, locked=2.14253095 USDT, free=100_016.35844933 USDT)], margins=[MarginBalance(initial=2.13942290 USDT, maintenance=0.00310805 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=08706d62-a1fb-41e8-9940-3a70b7998131).
**2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderFilled(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220803-001-semiauto_ua_rt_dd_market-16, venue_order_id=BINANCE-1-016, account_id=BINANCE-001, trade_id=BINANCE-1-018, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, order_side=BUY, order_type=MARKET, last_qty=375.8, last_px=2.662 USDT, commission=0.40015184 USDT, liquidity_side=TAKER, ts_event=1659490006778000000).**
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.Portfolio: GTCUSDT-PERP.BINANCE net_position=374.7
2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.50098028 USDT, locked=3.19813625 USDT, free=100_015.30284403 USDT)], margins=[MarginBalance(initial=2.13942290 USDT, maintenance=1.05871335 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=6118892e-b37d-4c7a-92d9-e0f578a07273).
**2022-08-03T01:26:46.778000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] PositionChanged(instrument_id=GTCUSDT-PERP.BINANCE, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, account_id=BINANCE-001, opening_order_id=O-20220802-001-semiauto_ua_rt_dd_market-9, closing_order_id=None, entry=BUY, side=LONG, net_qty=374.7, quantity=374.7, 
peak_qty=374.7, currency=USDT, avg_px_open=2.7810247713884997, avg_px_close=2.750810597726439, realized_return=0.01086, realized_pnl=38.88777936 USDT, unrealized_pnl=-44.59858184 USDT, ts_opened=1659459164621000000, ts_last=1659490006778000000, ts_closed=0, duration_ns=0).**
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.49975664 USDT, locked=3.19192016 USDT, free=100_015.30783648 USDT)], margins=[MarginBalance(initial=2.13320681 USDT, maintenance=1.05871335 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=77e41f0f-cd3b-4bd0-a2a5-c9cbb50c3b4b).
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] OrderFilled(instrument_id=GTCUSDT-PERP.BINANCE, client_order_id=O-20220802-001-semiauto_ua_rt_dd_market-15, venue_order_id=BINANCE-1-015, account_id=BINANCE-001, trade_id=BINANCE-1-019, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, order_side=BUY, order_type=STOP_MARKET, last_qty=1.1, last_px=2.781 USDT, commission=0.00122364 USDT, liquidity_side=TAKER, ts_event=1659497372170000000).
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.Portfolio: GTCUSDT-PERP.BINANCE net_position=375.8
2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.Portfolio: Updated AccountState(account_id=BINANCE-001, account_type=MARGIN, base_currency=USDT, is_reported=False, balances=[AccountBalance(total=100_018.49975664 USDT, locked=3.19502821 USDT, free=100_015.30472843 USDT)], margins=[MarginBalance(initial=2.13320681 USDT, maintenance=1.06182140 USDT, instrument_id=GTCUSDT-PERP.BINANCE)], event_id=4dfe9ed2-4d93-437f-8eae-ed5162505369).
**2022-08-03T03:29:32.170000000Z [INF] BACKTESTER-001.UAM-semiauto_ua_rt_dd_market: <--[EVT] PositionChanged(instrument_id=GTCUSDT-PERP.BINANCE, position_id=GTCUSDT-PERP.BINANCE-UAM-semiauto_ua_rt_dd_market, account_id=BINANCE-001, opening_order_id=O-20220802-001-semiauto_ua_rt_dd_market-9, closing_order_id=None, entry=BUY, side=LONG, net_qty=375.8, quantity=375.8, 
peak_qty=375.8, currency=USDT, avg_px_open=2.781024698880444, avg_px_close=2.750810597726439, realized_return=0.01086, realized_pnl=38.88655572 USDT, unrealized_pnl=-0.00928184 USDT, ts_opened=1659459164621000000, ts_last=1659497372170000000, ts_closed=0, duration_ns=0).**
